### PR TITLE
Add a Dockerfile for Windows

### PR DESCRIPTION
**Pull request by benmoss**
@fujimotos can you take a look?

I built this on top of #1656 so that I didn't have to bother with installing OpenSSL. It shouldn't be too hard to do it with `scoop` though in the same place we're installing `winflexbison`.

#960

### CODE REVIEW

**Comment by fujimotos** [comment]
I think I can take a look into this patch in a couple of days.
I need to set up Docker for Windows on my environment first. WFM.

**Comment by benmoss** [comment]
We can look into building the image on Appveyor too, I experimented with this on https://github.com/benmoss/fluent-bit/tree/appveyor and have a build running here https://ci.appveyor.com/project/benmoss/fluent-bit

**Comment by fujimotos** [comment]
OK. I spent an hour this evening tracing your Dockerfile line by line, and
it mostly seems good to me. I left some review comment on things I noticed.

That said, I have not really tested this Dockerfile on my machine yet.
I am planning to take time to set up Docker for Windows next week and
see if it actually works, so please wait a bit more.

File: Dockerfile.windows [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                     -D FLB_WITHOUT_flb-rt-out_elasticsearch=On `
+                     -D FLB_WITHOUT_flb-rt-out_td=On `
+                     -D FLB_WITHOUT_flb-rt-out_forward=On `
+                     -D FLB_WITHOUT_flb-rt-in_disk=On `
+                     -D FLB_WITHOUT_flb-rt-in_proc=On `
+                     -D FLB_WITHOUT_flb-it-parser=On `
+                     -D FLB_WITHOUT_flb-it-unit_sizes=On `
+                     -D FLB_WITHOUT_flb-it-network=On `
+                     -D FLB_WITHOUT_flb-it-pack=On `
+                     ../
```
Comment (fujimotos) [inline Dockerfile.windows]: I think we don't need to build tests on Docker (`FLB_TESTS_INTERNAL=On`).
This means that you can simplify this build command to:

```
RUN powershell -Command cmake -G "$ENV:msvc" -DCMAKE_BUILD_TYPE="$ENV:configuration" ..
```

**Comment by edsiper** [comment]
including @paulbouwer on this, we recently discussed Windows container.

Paul, what do you think?

**Comment by fujimotos** [comment]
> That said, I have not really tested this Dockerfile on my machine yet.
> I am planning to take time to set up Docker for Windows next week and
> see if it actually works, so please wait a bit more.

@benmoss  @edsiper  OK. I spent this evening experimenting this
patch on Windows Server 2019 and confirmed that this Dockerfile
works good.

Here is some screenshot:

docker

**Comment by paulbouwer** [comment]
@edsiper @fujimotos - I have pulled together all of bits from the POC we recently did and I'll add it as a PR for discussion. I feel that there are bits of it that would probably belong in the docs:

- [Fluent Bit Manual - Installation > Kubernetes](https://docs.fluentbit.io/manual/installation/kubernetes)
- [Fluent Bit Manual - Installation > Windows](https://docs.fluentbit.io/manual/installation/windows)

And bits that belong in this repo.

I had built out a PowerShell version of [appveyor-download](https://github.com/fluent/fluent-bit/blob/master/ci/appveyor-download) to allow use of the Windows Server Core base image (as is) to be used to download the correct artifact. I see that the Windows artifacts are now available at http://fluentbit.io/releases, which should make the Dockerfile simpler. I'll include the `appveyor-download.ps1` script since it may be useful for something else.

- http://fluentbit.io/releases/1.3/td-agent-bit-1.3.3-win64.zip

**Comment by fujimotos** [comment]
@benmoss Sorry for some delay, 

The thing is, I'd like to merge your work along with #1801, since I believe 
your version is useful for development purpose, especially when building a
custom image for Windows.

So can you apply some updates to this PR:

 1. Rename `Dockerfile.windows` to `Dockerfile.windows.devel` 
 2. Add sign-off lines to commits to clear away DCO warnings.

... so that we can incorporate your work?

**Comment by benmoss** [comment]
Yup, apologies for my delay too, I've been on holiday for the past few weeks. Will look into this soon.

**Comment by benmoss** [comment]
@fujimotos should be mergeable now, looks like just some random flake on Travis

**Comment by fujimotos** [comment]
@benmoss Thank you.

For what it's worth, I've tested this patch yesterday on my Development
environment and it worked. So I think this patch is mergeable now.

**Comment by edsiper** [comment]
is the final name of the Dockerfile correct: ```Dockerfile.windows.devel ``` ?

**Comment by benmoss** [comment]
That's what @fujimotos asked for

**Comment by edsiper** [comment]
@fujimotos why Dockerfile.windows.devel ?

**Comment by fujimotos** [comment]
> why Dockerfile.windows.devel ?

@edsiper To avoid conflict with #1801, so that we can safely merge both PRs
into master.

To put it short, #1760 and #1801 use different approaches to build images.

* #1801 *(Dockerfile.windows)*
  * Uses official prebuild `.exe` distributed on fluentbit.io.
  * Fast and lightweight.
  * Should be useful for production deployment.
* #1760 *(Dockerfile.windows.devel)*
  * Builds an image from scratch by compiling the source.
  * Requires more system resource, but quite useful for development.
  * Also useful for users who want to deploy a custom build.

So both versions are useful for different situations; Since both PRs cannot
uses the same file name "Dockerfile.windows", I suggested @benmoss to
rename this one to "Dockerfile.windows.devel".

**Comment by edsiper** [comment]
#1760 and #1801 can both be integrated into the CI easily ?

**Comment by fujimotos** [comment]
> #1760 and #1801 can both be integrated into the CI easily ?

@edsiper #1760 is a better to integrate with CI. #1801 needs a Windows binary
released on fluentbit.io first, so might not be useful for CI integration.

Just for your information, I'm planning to add some fixup patch on top of #1801 
and #1760 to clean up various small things (Make the build definition looks more
homogeneous each other; Add documentation; etc etc ) 

If you want a better filename/file layout, I think we can do that my subsequent PRs.
(i.e. I'd like to aid this fine contribution into master, and fix small things by ourselves)

**Comment by fujimotos** [comment]
@benmoss Thank you for taking your time and contribute this patch.
It will be included in the next v1.4.0 release. I'd very appreciate it.

**Comment by fujimotos** [comment]
@edsiper Thank you!
 
@benmoss Thank you for taking your time and contribute this patch.
It will be included in the next v1.4.0 release. I appreciate it.

**Pull request merged**